File: kaji/sdk/src/integrations/registry/echo/echo.py
"""Echo integration. The simplest possible Kaji integration.

Two pure functions, no auth, no network. Installed by `kaji add echo`.
"""

from __future__ import annotations

import kaji


def _message_parameters() -> dict[str, object]:
    return {
        "$schema": "https://json-schema.org/draft/2020-12/schema",
        "type": "object",
        "properties": {"message": {"type": "string"}},
        "required": ["message"],
        "additionalProperties": False,
    }


@kaji.function_tool(
    description="Return the input string unchanged.",
    parameters=_message_parameters(),
    risk="read",
    namespace="echo",
    parallel_safe=False,
)
async def say(message: str) -> dict:
    return {"message": message}


@kaji.function_tool(
    description="Return the input string uppercased.",
    parameters=_message_parameters(),
    risk="read",
    namespace="echo",
    parallel_safe=False,
)
async def shout(message: str) -> dict:
    return {"message": message.upper()}


tools = (say, shout)

File: kaji/src/kaji/integrations/registry/echo/echo.ts
// This is YOUR echo integration. Edit it.
// Common customizations:
//   1. Trim the tools array to just what your agent needs
//   2. Tighten Zod schemas — make fields required if your agent should never miss them
//   3. Add helper tools your agent wants but the API doesn't have natively
// Updates: re-run `kaji add echo` to diff against the latest version we ship.

import { functionTool, Integration } from "kaji-sdk";
import * as z from "zod";

const messageParameters = z.strictObject({ message: z.string() });

export const say = functionTool(
  {
    name: "say",
    namespace: "echo",
    description: "Return the input string unchanged.",
    parameters: messageParameters,
    risk: "read",
    parallel_safe: false,
  },
  async ({ message }) => ({ message }),
);

export const shout = functionTool(
  {
    name: "shout",
    namespace: "echo",
    description: "Return the input string uppercased.",
    parameters: messageParameters,
    risk: "read",
    parallel_safe: false,
  },
  async ({ message }) => ({ message: message.toUpperCase() }),
);

/** Side-effect-free metadata source for registry ABI verification. */
export const tools = Object.freeze([say, shout] as const);

export class EchoIntegration extends Integration {
  readonly namespace = "echo";

  override tools() {
    return tools.map(
      (bound) => [bound.spec, bound.handler] as [typeof bound.spec, typeof bound.handler],
    );
  }
}

/** Construct Echo without executing a tool or reading runtime state. */
export function inspectIntegration(): EchoIntegration {
  return new EchoIntegration();
}
